### ISSUE TITLE
no-object-literal-type-assertion in object literals

### ISSUE THREAD

**Issue by pronebird** [palantir/tslint#4503]
### Bug Report

- __TSLint version__: 5.12.1
- __TypeScript version__: 3.x
- __Running TSLint via__: CLI

#### TypeScript code being linted

```ts
type Bar = { key: 'a' } | { key: 'b', extra: boolean }; 

const foo = {
   // no-object-literal-type-assertion  Type assertion on object literals is forbidden, use a type annotation instead.
   bar: { key: 'a' } as Bar, 
}
```

#### Actual behavior

`no-object-literal-type-assertion` is being raised where it's basically impossible to declare a type without using `as` expression or moving the value outside of the object literal.

#### Expected behavior

I don't know, but perhaps allow `as` in such scenario. But if the behaviour is expected, then the [documentation](https://palantir.github.io/tslint/rules/no-object-literal-type-assertion/) has to be updated with the proper example.

**Comment by JoshuaKGoldberg** [comment]
@pronebird In your example, why did you want to add the `as Bar`? In your code, `{ key: 'a' }` is a valid object of type `Bar`. It should actually have a better type than `Bar`, since TypeScript has inferred it's not the second/`'b'` form.

Is it that you specifically want `foo` to be of the `{ bar: Bar }`? If that's the case, how about declared `const foo: { bar: Bar } = ...`?

**Comment by pronebird** [comment]
@JoshuaKGoldberg alright so my use case was originally related to building a singleton object with a bunch of methods that I didn't want to formally describe in an interface because the object was not exported anywhere and was  only used within the same file, so TypeScript should be able to infer the types on its own.

However down the road, I've switched to using a class because, sadly, access modifiers are not supported in literal objects, i.e:

```ts
const Foo = {
  private bar() { // can't use "private"

  }
}
```

Besides, defining the types of fields in a literal object is not simple in TypeScript.

```ts
// Using literal
const StaticFoo = {
  bar: Bar = .. // syntax error apparently
  
  bar: ({ key: 'a' } as Bar) // OK
}

// Using class
class NonStaticFoo {
  public bar: Bar = ... // OK
}
```

**Comment by adidahiya** [comment]
closing for now, might consider reopening if there is more feedback
